app/src/main/java/io/mercury/coroutinesandbox/view/downloader/DownloadListActivity.kt

```diff
@@ -0,0 +1,131 @@
+class DownloadListActivity : ComponentActivity() {
+
+    @Inject
+    lateinit var downloadMovie: DownloadMovie
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        val model: DownloadListViewModel by viewModels()
+
+        setContent {
+            val composeState by model.feature.state.collectAsState(lifecycleScope.coroutineContext)
+            Root(composeState)
+        }
+
+        model.feature.load()
+    }
+
+    @Composable
+    fun Root(state: State) {
+        ThemedMaterial {
+            Box(
+                Modifier
+                    .background(MaterialTheme.colors.background)
+                    .fillMaxSize()
+                    .padding(16.dp)
+            ) {
+                when (state) {
+                    is Uninitialized,
+                    is Loading -> Text("Loading...")
+                    is Error -> Text("ERROR")
+                    is Loaded -> {
+                        Column(
+                            verticalArrangement = Arrangement.spacedBy(24.dp)
+                        ) {
+                            Column(
+                                modifier = Modifier
+                                    .weight(1f)
+                                    .background(Color.LightGray)
+                                    .fillMaxWidth()
+                            ) {
+                                Text("Downloading:")
+                                LazyColumn {
+                                    items(state.inProgressDownloads) { download ->
+                                        Text("${download.title} - ${download.downloadProgress}%")
+                                    }
+                                }
+                            }
+
+                            Column(
+                                modifier = Modifier
+                                    .weight(1f)
+                                    .background(Color.Cyan)
+                                    .fillMaxWidth()
+                            ) {
+                                Text("Finished:")
+                                LazyColumn {
+                                    items(state.finishedDownloads) { download ->
+                                        Text(download.title)
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+
+                Button(
+                    onClick = ::downloadSomething,
+                    modifier = Modifier.align(Alignment.BottomEnd)
+                ) {
+                    Text("Download Something")
+
+                }
+            }
+        }
+    }
+
+    private val downloadables = Stack<String>().also {
+        it.push("tt0068646")
+        it.push("tt0111161")
+        it.push("tt0108052")
+    }
+
+    private fun downloadSomething() {
+        MainScope().launch {
```

Comment:
Referencing / Creating a scope like this kinda scares me (maybe it shouldn't?). I usually prefer injecting scope or passing it as a parameter so the caller always has control of the scope. In other words, how would I test this function?

Reply:
Again, this was laziness. Passing the scope into the feature would be my preferred way to go.